numba/tests/test_types.py

```diff
@@ -576,5 +580,43 @@ def impl():
         jit_impl = jit(nopython=True)(impl)
         self.assertEqual(impl(), jit_impl())
 
+class TestContainers(TestCase):
+    def test_list_type(self):
```

Comment:
I think this testing is too involved and nothing actually needs compiling. If I recall correctly, the issue was that a `types.List` `__getitem__` call was using the default one inherited from `types.Type` metaclass. I think the minimum test coverage path is to simply check behaviour along the lines of:
```python
def test_list_type_getitem(self):
    for listty in (types.int64, types.Array(types.float64, 1, 'C')):
        l_int = types.List(listty)
        self.assertTrue(isinstance(l_int, types.List))
        self.assertTrue(isinstance(l_int[0], type(listty)))
```
which is asserting that `__getitem__` on the `list.Type` does indeed return the type of the items in the list.

Reply:
you're right, that is the minimum test that can be done. If you think it's enough, then I'll leave it at that. I created a more involved test to replicate the context in which the problem can show up. Admittedly, mostly because I don't know the entire path the type classes take between definition and use in `@overload` or `@generated_jit` and therefore I though it would be safer to test "end to end". Happy to remove if you think the simpler test is enough.

Reply:
I think this is sufficient given what the issue is. The type class instances are a result of successfully completing type inference whereby all the variables are assigned a concrete type. What the current tests are doing is running the compiler, and therefore type inference, to obtain types (and then doing some compilation based on those types later).